[docs] Add a tutorial for the pass manager (#3515)
dmlc/tvm: docs/dev/relay_pass_manager.rst

..  Licensed to the Apache Software Foundation (ASF) under one
    or more contributor license agreements.  See the NOTICE file
    distributed with this work for additional information
    regarding copyright ownership.  The ASF licenses this file
    to you under the Apache License, Version 2.0 (the
    "License"); you may not use this file except in compliance
    with the License.  You may obtain a copy of the License at

..    http://www.apache.org/licenses/LICENSE-2.0

..  Unless required by applicable law or agreed to in writing,
    software distributed under the License is distributed on an
    "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
    KIND, either express or implied.  See the License for the
    specific language governing permissions and limitations
    under the License.

Relay Pass Infra: the Pass Manager
==================================

Relay features a series of optimization passes which improve performance metrics

> of models such as mean inference, memory footprint, power consumption, or for

Reviewer: or power consumption for specific devices.

> specific devices. There are a suite of standard optimizations as well as machine

Reviewer 2: "or for specific devices" seems odd to read

learning specific optimizations including constant folding, dead code
elimination, operator layout alteration, and operator fusion, etc. Each of these
passes are structured as Relay-to-Relay transformations on the abstract syntax
tree (AST) using the analysis result collected during and/or before traversal.

On one hand, many existing production compilers, such as GCC and LLVM, employ
pass managers to effectively manage the execution of passes. Initially managing
passes is straightforward as the number of passes is small, but mature compilers
will contain hundreds of individual passes. Often external users will want to
have custom passes correctly scheduled without having to modify a single
handcrafted pass order.

> One the other hand, modern deep learning frameworks, such as Pytorch and MXNet

Reviewer 2: One -> On

Gluon, also have the tendency to enable pass manager style layer construction
scheme through `Sequential`_ and `Block`_, respectively. With such constructs,
these modern frameworks are able to conveniently add modules/layers to their
containers and build up neural network easily.

.. _Sequential: https://pytorch.org/docs/stable/nn.html?highlight=sequential#torch.nn.Sequential
.. _Block: https://mxnet.incubator.apache.org/_modules/mxnet/gluon/block.html

The design of the Relay pass manager is largely inspired by the the hierarchical
pass manager used in LLVM and the block style containers used in the popular
deep learning frameworks. The major goals of the pass manager include:

#) enabling better programmatic orchestration of optimizations. This allows
   users to flexibly customize and build their own optimization pipelines.

#) providing a user friendly way to debug optimization passes.

> #) alleviating developers's from manually and respectively resolving the

Reviewer 3: ```suggestion
#) alleviating developers from manually and respectively resolving the
```

level the pass will be enabled, and ``required`` represents the passes that are
required to execute a certain pass (see ``include/tvm/relay/transform.h`` for
more details).

.. code:: c++

    class PassInfoNode : public RelayNode {
      std::string name;
      int opt_level;
      std::vector<std::string> required;
    };

PassContext
^^^^^^^^^^^

``PassContext`` carries useful information for an optimization pass. For
example, it contains the error reporting system so optimization authors can
provide diagnostics about why an optimization fails. ``PassContext`` is also
designed to replace the old ``BuildConfig`` which was used to help users
configure the compilation options, including optimization level and
required/disabled passes, etc. 

This class is designed as a thread local so that users can conveniently use the
Python ``with`` syntax to perform optimizations under a certain configuration.
Examples will be provided later to show how we can use both C++ and Python APIs
to create a compilation pipeline using pass context.

.. code:: c++

    class PassContextNode : public RelayNode {
     public:
      ErrorReporter err_reporter;

>       int opt_level{2};

Reviewer 2: Why is opt_level here? Isn't defining it in PassInfoNode sufficient?

Author: yeah, each pass is associated with an opt_level. So you will know if a certain pass will be enabled when users provide a opt_level in the configuration context. This is different from what we usually have in a compiler.

      int fallback_device{static_cast<int>(kDLCPU)};
      tvm::Array<tvm::Expr> required_pass;
      tvm::Array<tvm::Expr> disabled_pass;
    };

.. code:: c++

    class PassContext : public NodeRef {
     public:
      TVM_DLL static PassContext Create();
      TVM_DLL static PassContext Current();
      /* Other fields are omitted. */
    
     private:
      // The entry of a pass context scope.
      TVM_DLL void EnterWithScope();
      // The exit of a pass context scope.
      TVM_DLL void ExitWithScope();
    
      // Classes to get the Python `with` like syntax.
      friend class tvm::With<PassContext>;
    };
   

Pass Constructs
^^^^^^^^^^^^^^^

The pass manager is designed in a hierarchical manner, and it could work at
various granularity of a Relay program. A pure virtual class ``PassNode`` is

> introduced to sever as the base of the different optimization passes. This class

Reviewer 2: sever -> serve

contains several virtual methods that will be implemented by the
subclasses at the module level, function level, or a sequence of passes..

.. code:: c++

    class PassNode : RelayNode {
      virtual PassInfo Info() const = 0;
      virtual Module operator()(const Module& mod
                                const PassContext& pass_ctx) const = 0;
    };

The functor shows how a pass will be realized, i.e. it always works on a Relay
module under a certain context. All passes are designed in a ``Module`` to
``Module`` manner. Therefore, optimizations governed by the pass manager will
always update the whole module.

Several subclasses are created to implement different types of optimization
passes, e.g. function-level passes, module-level passes, and sequential passes.
Each subclass itself could act as a pass manager. For instance, they could glob
the require passes and execute them or build a dependency graph based on the
given meta data. The full definition of them could be found in

> ``src/relay/pass/pass_manager.cc``

Reviewer 3: Could directly make this a link to the [source file](https://github.com/dmlc/tvm/blob/master/src/relay/pass/pass_manager.cc) using one of [these methods](http://docutils.sourceforge.net/docs/user/rst/quickref.html#hyperlink-targets)

Module-level Passes
^^^^^^^^^^^^^^^^^^^

> Module level passes are geared mainly for global and inter-procedural

Reviewer 2: What is the example of IPO in this context?

Author: we have inlining could be viewed as one although we don't really have multiple translation units.

typical passes in Relay that need the global picture of a module, such as to
a normal form conversion and lambda lifting, etc, fall in this set. At this
level, uses can even add and/or delete functions in a module.

.. code:: c++

    class ModulePassNode : PassNode {
      PassInfo pass_info;
      runtime::TypedPackedFunc<Module(Module, PassContext)> pass_func;
      Module operator()(const Module& mod, const PassContext& pass_ctx) const final;
      // Other members/methods are omitted
    };

``pass_info`` maintains the information needed by a module-level pass.
``pass_func`` sketches the real optimization. For example, we may need to
perform dead code elimination on the module. We could implement the algorithm in
the ``pass_func`` and let it run on a module. It will then remove the dead code
including the unused functions in the module. Note that this field is designed
as a packed function, which enables the implementation of the optimization in
both C++ and Python.

Function-level Passes
^^^^^^^^^^^^^^^^^^^^^

Function-level passes are used to implement various intra function level
optimizations for a given Relay module. It fetches one function at a time from

> the function list of a module for optimization and yields a rewritten Relay

Reviewer 2: If there is a link for relay module description, it will be good to add here.

common subexpression elimination and inference simplification, etc.

Note that the scope of passes at this level is a Relay function. Therefore, we
cannot add or delete a function through these passes as they are not aware of
the global information.

.. code:: c++
   
    class FunctionPassNode : PassNode {
      PassInfo pass_info;
      runtime::TypedPackedFunc<Function(Function, Module, PassContext)> pass_func;
      Module operator()(const Module& mod, const PassContext& pass_ctx) const final;
      bool SkipFunction(const Function& func) const;
      // Other members/methods are omitted...
    };

``pass_info`` is identical to what we just described in the module pass.
``pass_func`` takes a function for optimization, it also needs a module as we
may use it for reporting errors. A function could be annotated with
"SkipOptimization" so that it will be ignored during optimization.

Sequential Passes
^^^^^^^^^^^^^^^^^

SequentialPass is similar to Pytorch nn.Sequential that contains a host of
passes for execution. It also has a disabled list that contains the list of
disabled passes. For example, we may have a configuration which performs all

> passes at ``opt_level=3`` with some disabled passes using disabled_pass=xx

Reviewer 3: ```suggestion
passes at ``opt_level=3`` with some disabled passes using ``disabled_pass=xx``
```

> provided by ``PassContext``. Now using ``SequentailPass``, we could glob all

Reviewer 3: ```suggestion
provided by ``PassContext``. Now using ``SequentialPass``, we could glob all
```

passes at ``opt_level=3`` and exclude those in disabled.

.. code:: c++

    class SequentialPassNode : PassNode {
      PassInfo pass_info;
      // Passes need to be executed.
      Array<Pass> passes;
      // Passes are disabled during optimization
      std::vector<std::string> disabled;
      bool PassEnabled(const PassInfo& info) const;
      Module operator()(const Module& mod, const PassContext& pass_ctx) const final;
    };

> Only a few passes currently in Relay is put in this group. For example,

Reviewer 3: ```suggestion
Only a few passes currently in Relay are put in this group. For example,
```

``FoldScaleAxis`` requires to dispatch ``ForwardFoldScaleAxis`` and
``BackwardFoldScaleAxis`` internally. In addition, ``BackwardFoldScaleAxis`` is
recommended to be fulfilled first. This pass, hence, is an ideal candidate for

> ``SequentailPass``.

Reviewer 3: ```suggestion
``SequentialPass``.
```

The following code shows how individual passes in a sequential pass is invoked.
Essentially, we sequentially execute each pass in a sequential pass using the
order that they were appended to the pass list without the consideration of

> their orders. The pass phase ordering problem might be considered in the future.

Reviewer 3: "using the order that they were appended to the pass list without the consideration of their orders" sounds like a contradiction. maybe trim off the end of the sentence?

> their orders. The pass phase ordering problem might be considered in the future.

Reviewer 3: Is it relevant to bring up the pass phase ordering problem here?

Author: Removed

.. code:: c++

    Module SequentialNode::operator()(const Module& module,
                                      const PassContext& pass_ctx) const {
      Module mod = module;
      for (const Pass& pass : passes) {
        CHECK(pass.defined()) << "Found undefined pass for optimization.";
        const PassInfo& pass_info = pass->Info();
        if (!PassEnabled(pass_info))  continue;
        for (const auto& it : pass_info->required) {
          const auto* name = it.as<tvm::ir::StringImm>();
          CHECK(name);
          mod = GetPass(name->value)(mod, pass_ctx);
        }
        mod = pass(mod, pass_ctx);
      }
      return mod;
    }

Upon the invocation of a pass, we first check if this pass is enabled. This is
done by first checking if the pass is explicitly disabled by a user, followed by
inspecting if it is specified as a required pass by the user. If it is still

> indetermined that whether this pass is enabled, its ``opt_level`` will be

Reviewer 3: ```suggestion
undetermined whether this pass is enabled, its ``opt_level`` will be
```

checked. This pass will be enabled and therefore executed only when its

> optimization level not less than the configured optimization level in the pass

Reviewer 3: ```suggestion
optimization level is not less than the configured optimization level in the pass
```

> context.

Reviewer 2: So, opt_level will be ignored, if user just specifies that I want to run Pass A, correct? Just like gcc -opt_A works even if we dont provide -O3 etc.

Author: yes

passes. These helpers are also exposed to the Python frontend for users to
favorably use Python APIs to create a specific pass object.

.. code:: c++

    FunctionPass CreateFunctionPass(std::string name,
                                    int opt_level,
                                    PassFunc pass_func);

    ModulePass CreateModulePass(std::string name,
                                int opt_level,
                                PassFunc pass_func);
    
    SequentialPass CreateSequentialPass(std::string name,
                                        int opt_level,
                                        Array<Pass> passes,
                                        Array<tvm::Expr> disabled);

C++ Sequential Example
^^^^^^^^^^^^^^^^^^^^^^

Let's now take an example to illustrate how the pass manager works on
``SequentialPass``. For illustrative purpose, only a code snippet is provided.
First, we create a simple Relay program, ``y = f(x)``. Then, we build a module
based on the function. After creating the module, we instantiate a sequential
pass object which contains some standard Relay optimization passes, including

> typer inference, dead code elimination, common subexpression elimination, and

Reviewer 3: ```suggestion
type inference, dead code elimination, common subexpression elimination, and
```

example, users can directly apply const folding pass on a given module, ``mod
= transform::FoldConstant()(mod)``. However, it is users' responsibility to
execute the required passes explicitly.

Pass Registration
~~~~~~~~~~~~~~~~~

We've covered the concept of different level of passes and the context used for
compilation. It would be interesting to see how easily users can register
a pass.  Let's take const folding as an example. This pass has already been
implemented to fold constants in a Relay function (found in
``src/relay/pass/fold_constant.cc``).

An API was provided to perform the ``Expr`` to ``Expr`` transformation.

.. code:: c++

    Expr FoldConstant(const Expr& expr);

In order to register this pass to the pass manager, we first need to decide at
which level this pass will be performed. As const folding happens on individual
functions, we should intuitively create a ``FunctionPass`` for it through
``CreateFunctionPass``. The ``pass_func`` is returned as a packed function that
invokes the ``Expr`` to ``Expr`` API on each function in a Relay module. ``{}``
indicates that no prerequisite is required for this pass. Otherwise, the pass

> develop has to identify and list them.

Reviewer 3: ```suggestion
developer has to identify and list them.
```

Meanwhile, a pass API endpoint is registered with the name
``relay._transform.FoldConstant``. This pass, therefore, becomes an entry in the
registry that can be accessed by both C++ (e.g. the ``GetPass`` above) and
Python when needed.

.. code:: c++

    namespace transform {

    Pass FoldConstant() {
      runtime::TypedPackedFunc<Function(Function, Module, PassContext)> pass_func =
        [=](Function f, Module m, PassContext pc) {
          return Downcast<Function>(FoldConstant(f));
      };
      return CreateFunctionPass(pass_func, 2, "FoldConstant", {});
    }

    TVM_REGISTER_API("relay._transform.FoldConstant")
    .set_body_typed(FoldConstant);

    }  // namespace transform

To allow other C++ modules to apply this pass, we declare a free function in

> `include/tvm/relay/transform.h` as the following:

Reviewer 3: ```suggestion
``include/tvm/relay/transform.h`` as the following:
```

a certain scope.

.. code:: python

    @register_relay_node
    class PassContext(RelayNode):
        def __enter__(self):
            _transform.EnterPassContext(self)
            return self
    
        def __exit__(self, ptype, value, trace):
            _transform.ExitPassContext(self)
    
        @staticmethod
        def current():
            """Return the current pass context."""
            return _transform.GetCurrentPassContext()

A ``PassContext`` object can be instantiated through the ``build_config`` API
which was used by Relay to configure the compilation options, including the
optimization level, fallback device for heterogeneous execution, and
required/disabled passes.

Pass Objects
^^^^^^^^^^^^

``Pass`` is the base class of all pass objects. All methods here are just simple
wrappers that were implemented in the backend. They are defined for users to
conveniently interact with the base class in Python. Only a functor is defined
to invoke the execution of a pass.

.. code:: python

    @register_relay_node
    class Pass(RelayNode):
       def __call__(self, mod):
           return _transform.RunPass(self, mod)

> Some auxiliary APIs are provided to enable to easy creation of passes from

Reviewer 3: ```suggestion
Some auxiliary APIs are provided to enable easy creation of passes from
```

> the Python frontend and let the pass manager to control the execution. For

Reviewer 3: ```suggestion
the Python frontend and to let the pass manager control the execution. For
```

example, ``module_pass``, ``function_pass``, and ``sequential`` are provided to
users so that they can customize their own pass or pass pipeline.

> For all the passes that implemented in the C++ backend, we provide

Reviewer 3: ```suggestion
For all the passes that are implemented in the C++ backend, we provide
```

a corresponding Python API in ``python/tvm/relay/transform.py``. For instance,
const folding has a Python API like the following:

.. code:: python

    def FoldConstant():
        return _transform.FoldConstant()

Users can build a pass through decoration like the following:

.. code:: python

    @relay.transform.module_pass(opt_level=2)
    def transform(mod, ctx):
       tp = relay.TensorType((10,), "float32")
       x = relay.var("x", tp)
       gv = relay.GlobalVar("abs")
       func = relay.Function([x], relay.abs(x))
       new_mod = relay.Module({gv: func})
       new_mod.update(mod)
       return new_mod

   module_pass = transform
   assert isinstance(module_pass, transform.ModulePass)
   assert module_pass.info.opt_level == 2

The ``transform`` function here adds an ``abs`` function to the input module,
but it could be any customized optimizations at the module level. After

> creating this `module_pass`, users can apply it on any Relay module. For

Reviewer 3: ```suggestion
creating this ``module_pass``, users can apply it on any Relay module. For
```

example, we can build an empty module and apply this pass to add an ``abs``
function.

.. code:: python

    mod = relay.Module()
    mod = module_pass(mod)

> Correspondingly, we also offer such functionality for `function_pass`. For

Reviewer 3: ```suggestion
Correspondingly, we also offer such functionality for ``function_pass``. For
```

instance, an example function-level pass could be written as the following:

.. code:: python

    @relay.transform.function_pass(opt_level=1)
    class TestReplaceFunc:
       def __init__(self, new_func):
          self.new_func = new_func
          def transform_function(self, func, mod, ctx):
             # Just for demo purposes
             # Transform func to new_func
             return self.new_func

    x = relay.var("x", shape=(10, 20))
    f1 = relay.Function([x], x)
    f2 = relay.Function([x], relay.log(x))
    # fpass is now a special pass that replaces every
    # function to f1
    fpass = TestReplaceFunc(f1)
    # Now every function in input_mod is replaced by f1
    res_mod = fpass(input_mod)

Alternatively, users can also directly register a pass without using the

> decroators and then invoke it. Let's use ``Sequential`` to demo this scenario.

Reviewer 3: ```suggestion
decorators and then invoke it. Let's use ``Sequential`` to demo this scenario.
```

Python Sequential Example
^^^^^^^^^^^^^^^^^^^^^^^^^

This example not only illustrates how users can directly create a sequential
pass using Python APIs (this could be applied to module- and function- level

> passes as well), but only explains we can build an optimization pipeline using

Reviewer 3: ```suggestion
passes as well), but also explains how we can build an optimization pipeline using
```

``Sequential`` associated with other types of passes.

.. code:: python

    # Create a simple Relay program.
    shape = (1, 2, 3)
    c_data = np.array(shape).astype("float32")
    tp = relay.TensorType(shape, "float32")
    c = relay.const(c_data)
    x = relay.var("x", tp)
    y = relay.add(c, c)
    y = relay.multiply(y, relay.const(2, "float32"))
    y = relay.add(x, y)
    z = relay.add(y, c)
    z1 = relay.add(y, c)
    z2 = relay.add(z, z1)
    func = relay.Function([x], z2)
  
    # Customize the optimization pipeline. 
    seq = _transform.Sequential([
        relay.transform.InferType(),
        relay.transform.FoldConstant(),
        relay.transform.EliminateCommonSubexpr(),
        relay.transform.AlterOpLayout()
    ])
  
    # Create a module to perform optimizations.
    mod = relay.Module({"main": func})
    

>     # User can disable any passes that their don't want to execute by providing

Reviewer 3: ```suggestion
    # User can disable any passes that they don't want to execute by providing
```
